### PR TITLE
Migrate getDataDirectory from libkiwix to kiwix-desktop

### PR DESCRIPTION
**Pull request by sgourdas**
This migrates the `getDataDirectory()` function from libkiwix, after some cleanups.
 - Also removes `KIWIX_DATA_DIR`

Fixes kiwix/libkiwix#1087 
Fixes kiwix/libkiwix#1085

Blocked by https://github.com/kiwix/libkiwix/pull/1107

### CODE REVIEW

File: src/downloadmanagement.h
Comment (veloman-yunkan) [inline src/downloadmanagement.h:211]: I don't think that this function should be declared in this header file, but if there is no better place then it should at least be moved to before the `#endif`

File: src/kiwixapp.h
Comment (veloman-yunkan) [inline src/kiwixapp.h:13]: Better include in the .cpp file

File: src/downloadmanagement.h
Comment (veloman-yunkan) [reply src/downloadmanagement.h:211]: I think that `settingsmanager.h/cpp` is a more appropriate home for this function's declaration and definition.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/kiwixapp.h
Comment (sgourdas) [reply src/kiwixapp.h:13]: Just for future reference, how should I decide where to include each header file?

Comment (veloman-yunkan) [reply src/kiwixapp.h]: Try to minimize dependencies (and don't introduce false dependencies). "kiwixapp.h" doesn't depend on "downloadmanagement.h" (the former doesn't use anything declared in the latter). By including "downloadmanagement.h" in "kiwixapp.h" you grow the list of formal dependencies of "downloadmanagement.h". Then if anything changes in "downloadmanagement.h" more code will be recompiled. If you include "downloadmanagement.h" only in the .cpp files that use stuff from that header, then recompilation in case of touching it will be faster.

**Comment by kelson42** [comment]
@sgourdas @veloman-yunkan What is the status here, it seems it needs a new review?

**Comment by sgourdas** [comment]
@kelson42 yes, ready for a new review

File: src/settingsmanager.cpp
Comment (mgautierfr) [inline src/settingsmanager.cpp:45]: I wonder if we could remove this and replace it with Qt function (writableLocation)[`https://doc.qt.io/qt-6/qstandardpaths.html#writableLocation`] (or (standardLocation)[https://doc.qt.io/qt-6/qstandardpaths.html#standardLocations], depending if we want to read data (configuration, zim files) or write it)

`QStandardPaths::AppDataLocation` seems to return what the commented lines are doing (at least on Windows and Linux)

**Comment by veloman-yunkan** [comment]
> @sgourdas @veloman-yunkan What is the status here, it seems it needs a new review?

@kelson42 I will review it as soon as kiwix/libkiwix#1107 is merged

File: src/downloadmanagement.cpp
Comment (veloman-yunkan) [inline src/downloadmanagement.cpp:105]: Undo this change. Call `getDataDirectory()` inside this function

**Comment by mgautierfr** [comment]
> I will review it as soon as https://github.com/kiwix/libkiwix/pull/1107 is merged

I have revert the merge of #1107 as it breaks the nightly CI until this PR is merged.
Please review this PR as if #1107 is merged and I will revert the revert PR (https://github.com/kiwix/libkiwix/pull/1110)

File: src/settingsmanager.cpp
Comment (sgourdas) [reply src/settingsmanager.cpp:45]: Good idea. I looked into it and the only issue I see is the intermediate differences between Qt and us currently For example, I have the resolutions below: 

## Windows
Current:

1. `C:\Users\<USER>\AppData\Roaming\kiwix`
2. `C:\Users\<USER>\kiwix`
3. `<APPDIR>`

AppDataLocation:

1. `C:/Users/<USER>/AppData/Roaming/kiwix`
2. `C:/ProgramData/kiwix`
4. `<APPDIR>`
5. `<APPDIR>/data`
6. `<APPDIR>/data/kiwix`

## Rest
Current:

1. `~/.local/share/kiwix`
2. `<APPDIR>`

AppDataLocation:

1. `~/.local/share/<APPNAME>`
3. `/usr/local/share/<APPNAME>`
4. `/usr/share/<APPNAME>`

How should we go about this?

**Comment by kelson42** [comment]
@veloman-yunkan We need a review now, before merging the libkiwix part, otherwise in the meantime the whole kiwix desktop CI will be broken.

**Comment by sgourdas** [comment]
On this I am waiting for a reply to [my comment](https://github.com/kiwix/kiwix-desktop/pull/1169/files#r1714974074) on if it is a good way to proceed.

**Comment by sgourdas** [comment]
I have made a suggestion commit to showcase the new implementation. Using `writableLocation` we get only the first valid path from the possible list of `AppDataLocation` and create it if not already there. This should work as intended.

**Comment by kelson42** [comment]
@veloman-yunkan We need a new review pass I guess.

**Review by veloman-yunkan — COMMENTED** [review]
There are too many commits in the PR history

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**
